adapters/gst/gst_plugins/python/savant_parse_bin.py

```diff
@@ -0,0 +1,104 @@
+"""SavantParseBin element."""
+import inspect
+from typing import Optional
+
+from savant.gstreamer import GObject, Gst  # noqa:F401
+from savant.gstreamer.codecs import CODEC_BY_CAPS_NAME, Codec
+from savant.gstreamer.utils import on_pad_event, propagate_gst_error
+from savant.utils.logging import LoggerMixin
+
+SINK_PAD_TEMPLATE = Gst.PadTemplate.new(
+    'sink',
+    Gst.PadDirection.SINK,
+    Gst.PadPresence.ALWAYS,
+    Gst.Caps.from_string(';'.join(x.value.caps_name for x in Codec)),
+)
+SRC_PAD_TEMPLATE = Gst.PadTemplate.new(
+    'src',
+    Gst.PadDirection.SRC,
+    Gst.PadPresence.ALWAYS,
+    Gst.Caps.from_string(';'.join(x.value.caps_with_params for x in Codec)),
+)
+
+
+class SavantParseBin(LoggerMixin, Gst.Bin):
+    """Applies an appropriate parser to the stream."""
+
+    GST_PLUGIN_NAME = 'savant_parse_bin'
+
+    __gstmetadata__ = (
+        'Savant parse bin',
+        'Generic/Bin/Parser',
+        'Applies an appropriate parser to the stream.',
+        'Pavel Tomskikh <tomskih_pa@bw-sw.com>',
+    )
+
+    __gsttemplates__ = (SINK_PAD_TEMPLATE, SRC_PAD_TEMPLATE)
+
+    __gproperties__ = {}
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._parser: Optional[Gst.Element] = None
+        self._sink_pad: Gst.Pad = Gst.GhostPad.new_no_target_from_template(
+            'sink', SINK_PAD_TEMPLATE
+        )
+        self._src_pad: Gst.Pad = Gst.GhostPad.new_no_target_from_template(
+            'src', SRC_PAD_TEMPLATE
+        )
+        self.add_pad(self._sink_pad)
+        self.add_pad(self._src_pad)
+        self._sink_pad.set_active(True)
+        self._src_pad.set_active(True)
+
+        self._sink_pad.add_probe(
+            Gst.PadProbeType.EVENT_DOWNSTREAM,
+            on_pad_event,
+            {Gst.EventType.CAPS: self.on_sink_pad_caps},
+        )
+
+    def on_sink_pad_caps(self, pad: Gst.Pad, event: Gst.Event):
+        caps: Gst.Caps = event.parse_caps()
+        self.logger.debug('Got caps %r. Trying to find parser.', caps.to_string())
+        try:
+            codec = CODEC_BY_CAPS_NAME[caps[0].get_name()]
+        except KeyError:
+            error = f'Cannot found parser for caps {caps.to_string()!r}.'
```

Comment:
Is it not enough to set the correct caps for sink/source templates? 

L14 `Gst.Caps.from_string(';'.join(x.value.caps_name for x in Codec))`

Reply:
Probably that should be enough. I'll check it.

Reply:
Yes, it's enough to set caps for sink. Removed check.